config.org
#+TITLE: Lilahamstern's Doom Emacs Config
#+AUTHOR: Leo Ronnebro (The Nerdy Hamster)
#+STARTUP: showstars

* TABLE OF CONTENTS :toc:
- [[#identity][Identity]]
- [[#gui][GUI]]
  - [[#theme][Theme]]
  - [[#lines][Lines]]
  - [[#fonts][Fonts]]
- [[#custom][Custom]]
  - [[#fullscreen-startup][Fullscreen startup]]
  - [[#remote-connections][Remote Connections]]
- [[#langugaes][Langugaes]]
  - [[#c][C#]]
- [[#packages][Packages]]
  - [[#centaur-tabs][Centaur Tabs]]
  - [[#which-key][Which key]]
  - [[#magit][Magit]]
  - [[#org][Org]]
  - [[#projectile][Projectile]]
  - [[#elcord][Elcord]]
  - [[#mode-line][Mode-line]]
  - [[#wttrin-weather][Wttrin (Weather)]]
  - [[#selectric][Selectric]]
  - [[#treemacs][Treemacs]]
  - [[#keycast][Keycast]]

* Identity
#+begin_src emacs-lisp :tangle yes
(setq user-full-name "Leo Rönnebro"
      user-mail-address "leo.ronnebro@hamsterapps.net")
#+end_src

* GUI
** Theme
Default theme
#+begin_src emacs-lisp :tangle yes
(setq doom-theme 'doom-palenight)
#+end_src
*** Mode-line
Add battery status and current time to mode-line
#+begin_src emacs-lisp :tangle yes
(display-time-mode 1)                             
(unless (equal "Battery status not available"
               (battery))
  (display-battery-mode 1))
#+end_src
** Lines
Display line numbers
#+begin_src emacs-lisp :tangle yes
(setq display-line-numbers-type t)
(map! :leader
      :desc "Toggle truncate lines"
      "t t" #'toggle-truncate-lines)
#+end_src
** Fonts
#+begin_src emacs-lisp :tangle yes
(setq doom-font (font-spec :family "FiraCode Nerd Font Mono" :size 13)
      doom-variable-pitch-font (font-spec :family "sans" :size 13))
#+end_src
*** Ligatures
#+begin_src emacs-lisp :tangle yes
(use-package ligature
  :config
  ;; Enable the www ligature in every possible major mode
  (ligature-set-ligatures 't '("www"))

  ;; Enable ligatures in programming modes
  (ligature-set-ligatures 'prog-mode '("www" "**" "***" "**/" "*>" "*/" "\\\\" "\\\\\\" "{-" "::"
                                       ":::" ":=" "!!" "!=" "!==" "-}" "----" "-->" "->" "->>"
                                       "-<" "-<<" "-~" "#{" "#[" "##" "###" "####" "#(" "#?" "#_"
                                       "#_(" ".-" ".=" ".." "..<" "..." "?=" "??" ";;" "/*" "/**"
                                       "/=" "/==" "/>" "//" "///" "&&" "||" "||=" "|=" "|>" "^=" "$>"
                                       "++" "+++" "+>" "=:=" "==" "===" "==>" "=>" "=>>" "<="
                                       "=<<" "=/=" ">-" ">=" ">=>" ">>" ">>-" ">>=" ">>>" "<*"
                                       "<*>" "<|" "<|>" "<$" "<$>" "<!--" "<-" "<--" "<->" "<+"
                                       "<+>" "<=" "<==" "<=>" "<=<" "<>" "<<" "<<-" "<<=" "<<<"
                                       "<~" "<~~" "</" "</>" "~@" "~-" "~>" "~~" "~~>" "%%"))

  (global-ligature-mode 't))
#+end_src
* Custom
#+begin_src emacs-lisp :tangle yes
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…")               ; Unicode ellispis are nicer than "...", and also save /precious/ space

(global-subword-mode 1)                           ; Iterate through CamelCase words
#+end_src
** Fullscreen startup
#+begin_src emacs-lisp :tangle yes
(if (eq initial-window-system 'x)                 ; if started by emacs command or desktop file
    (toggle-frame-maximized)
  (toggle-frame-fullscreen))
#+end_src
** Remote Connections
#+begin_src emacs-lisp :tangle yes
(defun ssh-conn (host port)
  "Connect to a remote host by SSH."
  (interactive "sHost: \nsPort (default 22): ")
  (let* ((port (if (equal port "") "22" port))
         (switches (list host "-p" port)))
    (set-buffer (apply 'make-term "ssh" "ssh" nil switches))
    (term-mode)
    (term-char-mode)
    (switch-to-buffer "*ssh*")))
#+end_src

#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "SSH into custom server"
      "\\ d" #'ssh-conn())
#+end_src

* Langugaes
** C#
#+begin_src emacs-lisp :tangle yes
(eval-after-load
    'company
  '(add-to-list 'company-backends 'company-omnisharp))

(add-hook 'csharp-mode-hook #'company-mode)
;; (eval-after-load
;;     'company
;;   '(add-to-list 'company-backends #'company-omnisharp))

;; (defun custom-csharp-mode-setup ()
;;   (omnisharp-mode)
;;   (company-mode)

;;   (setq indent-tabs-mode t)
;;   (setq c-syntactic-indentation t)
;;   (c-set-style "ellemtel")
;;   (setq c-basic-offset 4)
;;   (setq truncate-lines t)
;;   (setq tab-width 4)
;;   (setq evil-shift-width 4)

;; ;;  (electric-pair-local-mode 1) ;; Emacs 25

;;   (local-set-key (kbd "C-c R") 'dotnet-run)
;;   (local-set-key (kbd "C-c r r") 'omnisharp-run-code-action-refactoring)
;;   (local-set-key (kbd "C-c f") 'omnisharp-code-format-entire-file)
;;   (local-set-key (kbd "C-c C-c") 'recompile))

;; (add-hook 'csharp-mode-hook 'custom-csharp-mode-setup t)
#+end_src

* Packages
** Centaur Tabs
#+begin_src emacs-lisp :tangle yes
(use-package centaur-tabs
  :init
  (setq centaur-tabs-enable-key-bindings t)
  :demand
  :config
  (centaur-tabs-mode t)
  (centaur-tabs-headline-match)
  (setq centaur-tabs-style "bar"
        centaur-tabs-set-icons t
        centaur-tabs-set-bar 'over
        centaur-tabs-set-close-button nil
        centaur-tabs-set-modified-marker t
        centaur-tabs-modified-marker "•"
        centaur-tabs-cycle-scope 'tabs))
(map! :leader
      :desc "Toggle centaur tabs on/off"
      "t c" #'centaur-tabs-local-mode)
#+end_src
** Which key
#+begin_src emacs-lisp :tangle yes
(setq which-key-idle-delay 0.5)
#+end_src
** Magit
#+begin_src emacs-lisp :tangle yes
(map! :leader
      :desc "Magit push"
      "g p" 'magit-push)
#+end_src
** Org
#+begin_src emacs-lisp :tangle yes
(after! org
  (require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-priority-faces '((65 :foreground "#e45649")
                             (66 :foreground "#da8548")
                             (67 :foreground "#0098dd"))
        org-directory "~/Documents/org/"
        org-ellipsis " ▼ "
        org-log-done 'time
        org-todo-keywords '((sequence "TODO(t)" "INPROGRESS(i)" "WAITING(w)" "|" "DONE(d)" "CANCELLED(c)"))
        org-todo-keyword-faces
        '(("TODO" :foreground "#7c7c75" :weight normal :underline t)
          ("WAITING" :foreground "#9f7efe" :weight normal :underline t)
          ("INPROGRESS" :foreground "#0098dd" :weight normal :underline t)
          ("DONE" :foreground "#50a14f" :weight normal :underline t)
          ("CANCELLED" :foreground "#ff6480" :weight normal :underline t))
        org-agenda-files (directory-files-recursively "~/Documents/org/" "\.org$")
        ))
#+end_src

*** Org Super Agenda
#+begin_src emacs-lisp :tangle yes
(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                   :time-grid t
                                   :scheduled today)
                                  (:name "Due Today"
                                   :deadline today)
                                  (:name "Important"
                                   :priority "A")
                                  (:name "Due soon"
                                   :deadline future)))
  :config
  (org-super-agenda-mode)
  )
#+end_src
*** Fancy priorities
#+begin_src emacs-lisp :tangle yes
(use-package org-fancy-priorities
  :ensure t
  :hook
  (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("⚡" "⬆" "⬇" "☕")))
#+end_src
** Projectile
#+begin_src emacs-lisp :tangle yes
(setq
 projectile-project-search-path '("~/code/*"))
#+end_src
** Elcord
#+begin_src emacs-lisp :tangle yes
(elcord-mode)
#+end_src
** Mode-line
#+begin_src emacs-lisp :tangle yes
(defun doom-modeline-conditional-buffer-encoding ()
  "We expect the encoding to be LF UTF-8, so only show the modeline when this is not the case"
  (setq-local doom-modeline-buffer-encoding
              (unless (or (eq buffer-file-coding-system 'utf-8-unix)
                          (eq buffer-file-coding-system 'utf-8)))))

(add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding)

(setq doom-modeline-env-version t
      doom-modeline-lsp t)

(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom")
#+end_src
** Wttrin (Weather)
#+begin_src emacs-lisp :tangle yes
(setq wttrin-default-cities '("Gothenburg"))
#+end_src
** Selectric
#+begin_src emacs-lisp :tangle yes
(require 'selectric-mode)
#+end_src
** Treemacs
*** Icons dired
#+begin_src emacs-lisp :tangle yes
(use-package treemacs-icons-dired
  :after treemacs dired
  :ensure t
  :config (treemacs-icons-dired-mode))
#+end_src
** Keycast
#+begin_src emacs-lisp :tangle yes
(use-package! keycast
  :commands keycast-mode
  :config
  (define-minor-mode keycast-mode
    "Show current command and its key binding in the mode line."
    :global t
    (if keycast-mode
        (progn
          (add-hook 'pre-command-hook 'keycast-mode-line-update t)
          (add-to-list 'global-mode-string '("" mode-line-keycast " ")))
      (remove-hook 'pre-command-hook 'keycast-mode-line-update)
      (setq global-mode-string (remove '("" mode-line-keycast " ") global-mode-string))))
  (custom-set-faces!
    '(keycast-command :inherit doom-modeline-debug
                      :height 0.9)
    '(keycast-key :inherit custom-modified
                  :height 1.1
                  :weight bold)))
#+end_src


config.el
(setq user-full-name "Leo Rönnebro"
      user-mail-address "leo.ronnebro@hamsterapps.net")

(setq doom-theme 'doom-palenight)

(display-time-mode 1)                             
(unless (equal "Battery status not available"
               (battery))
  (display-battery-mode 1))

(setq display-line-numbers-type t)
(map! :leader
      :desc "Toggle truncate lines"
      "t t" #'toggle-truncate-lines)

(setq doom-font (font-spec :family "FiraCode Nerd Font Mono" :size 13)
      doom-variable-pitch-font (font-spec :family "sans" :size 13))

(use-package ligature
  :config
  ;; Enable the www ligature in every possible major mode
  (ligature-set-ligatures 't '("www"))

  ;; Enable ligatures in programming modes
  (ligature-set-ligatures 'prog-mode '("www" "**" "***" "**/" "*>" "*/" "\\\\" "\\\\\\" "{-" "::"
                                       ":::" ":=" "!!" "!=" "!==" "-}" "----" "-->" "->" "->>"
                                       "-<" "-<<" "-~" "#{" "#[" "##" "###" "####" "#(" "#?" "#_"
                                       "#_(" ".-" ".=" ".." "..<" "..." "?=" "??" ";;" "/*" "/**"
                                       "/=" "/==" "/>" "//" "///" "&&" "||" "||=" "|=" "|>" "^=" "$>"
                                       "++" "+++" "+>" "=:=" "==" "===" "==>" "=>" "=>>" "<="
                                       "=<<" "=/=" ">-" ">=" ">=>" ">>" ">>-" ">>=" ">>>" "<*"
                                       "<*>" "<|" "<|>" "<$" "<$>" "<!--" "<-" "<--" "<->" "<+"
                                       "<+>" "<=" "<==" "<=>" "<=<" "<>" "<<" "<<-" "<<=" "<<<"
                                       "<~" "<~~" "</" "</>" "~@" "~-" "~>" "~~" "~~>" "%%"))

  (global-ligature-mode 't))

(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      truncate-string-ellipsis "…")               ; Unicode ellispis are nicer than "...", and also save /precious/ space

(global-subword-mode 1)                           ; Iterate through CamelCase words

(if (eq initial-window-system 'x)                 ; if started by emacs command or desktop file
    (toggle-frame-maximized)
  (toggle-frame-fullscreen))

(defun ssh-conn (host port)
  "Connect to a remote host by SSH."
  (interactive "sHost: \nsPort (default 22): ")
  (let* ((port (if (equal port "") "22" port))
         (switches (list host "-p" port)))
    (set-buffer (apply 'make-term "ssh" "ssh" nil switches))
    (term-mode)
    (term-char-mode)
    (switch-to-buffer "*ssh*")))

(map! :leader
      :desc "SSH into custom server"
      "\\ d" #'ssh-conn())

(eval-after-load
    'company
  '(add-to-list 'company-backends 'company-omnisharp))

(add-hook 'csharp-mode-hook #'company-mode)
;; (eval-after-load
;;     'company
;;   '(add-to-list 'company-backends #'company-omnisharp))

;; (defun custom-csharp-mode-setup ()
;;   (omnisharp-mode)
;;   (company-mode)

;;   (setq indent-tabs-mode t)
;;   (setq c-syntactic-indentation t)
;;   (c-set-style "ellemtel")
;;   (setq c-basic-offset 4)
;;   (setq truncate-lines t)
;;   (setq tab-width 4)
;;   (setq evil-shift-width 4)

;; ;;  (electric-pair-local-mode 1) ;; Emacs 25

;;   (local-set-key (kbd "C-c R") 'dotnet-run)
;;   (local-set-key (kbd "C-c r r") 'omnisharp-run-code-action-refactoring)
;;   (local-set-key (kbd "C-c f") 'omnisharp-code-format-entire-file)
;;   (local-set-key (kbd "C-c C-c") 'recompile))

;; (add-hook 'csharp-mode-hook 'custom-csharp-mode-setup t)

(use-package centaur-tabs
  :init
  (setq centaur-tabs-enable-key-bindings t)
  :demand
  :config
  (centaur-tabs-mode t)
  (centaur-tabs-headline-match)
  (setq centaur-tabs-style "bar"
        centaur-tabs-set-icons t
        centaur-tabs-set-bar 'over
        centaur-tabs-set-close-button nil
        centaur-tabs-set-modified-marker t
        centaur-tabs-modified-marker "•"
        centaur-tabs-cycle-scope 'tabs))
(map! :leader
      :desc "Toggle centaur tabs on/off"
      "t c" #'centaur-tabs-local-mode)

(setq which-key-idle-delay 0.5)

(map! :leader
      :desc "Magit push"
      "g p" 'magit-push)

(after! org
  (require 'org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-priority-faces '((65 :foreground "#e45649")
                             (66 :foreground "#da8548")
                             (67 :foreground "#0098dd"))
        org-directory "~/Documents/org/"
        org-ellipsis " ▼ "
        org-log-done 'time
        org-todo-keywords '((sequence "TODO(t)" "INPROGRESS(i)" "WAITING(w)" "|" "DONE(d)" "CANCELLED(c)"))
        org-todo-keyword-faces
        '(("TODO" :foreground "#7c7c75" :weight normal :underline t)
          ("WAITING" :foreground "#9f7efe" :weight normal :underline t)
          ("INPROGRESS" :foreground "#0098dd" :weight normal :underline t)
          ("DONE" :foreground "#50a14f" :weight normal :underline t)
          ("CANCELLED" :foreground "#ff6480" :weight normal :underline t))
        org-agenda-files (directory-files-recursively "~/Documents/org/" "\.org$")
        ))

(use-package! org-super-agenda
  :after org-agenda
  :init
  (setq org-super-agenda-groups '((:name "Today"
                                   :time-grid t
                                   :scheduled today)
                                  (:name "Due Today"
                                   :deadline today)
                                  (:name "Important"
                                   :priority "A")
                                  (:name "Due soon"
                                   :deadline future)))
  :config
  (org-super-agenda-mode)
  )

(use-package org-fancy-priorities
  :ensure t
  :hook
  (org-mode . org-fancy-priorities-mode)
  :config
  (setq org-fancy-priorities-list '("⚡" "⬆" "⬇" "☕")))

(setq
 projectile-project-search-path '("~/code/*"))

(elcord-mode)

(defun doom-modeline-conditional-buffer-encoding ()
  "We expect the encoding to be LF UTF-8, so only show the modeline when this is not the case"
  (setq-local doom-modeline-buffer-encoding
              (unless (or (eq buffer-file-coding-system 'utf-8-unix)
                          (eq buffer-file-coding-system 'utf-8)))))

(add-hook 'after-change-major-mode-hook #'doom-modeline-conditional-buffer-encoding)

(setq doom-modeline-env-version t
      doom-modeline-lsp t)

(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom")

(setq wttrin-default-cities '("Gothenburg"))

(require 'selectric-mode)

(use-package treemacs-icons-dired
  :after treemacs dired
  :ensure t
  :config (treemacs-icons-dired-mode))

(use-package! keycast
  :commands keycast-mode
  :config
  (define-minor-mode keycast-mode
    "Show current command and its key binding in the mode line."
    :global t
    (if keycast-mode
        (progn
          (add-hook 'pre-command-hook 'keycast-mode-line-update t)
          (add-to-list 'global-mode-string '("" mode-line-keycast " ")))
      (remove-hook 'pre-command-hook 'keycast-mode-line-update)
      (setq global-mode-string (remove '("" mode-line-keycast " ") global-mode-string))))
  (custom-set-faces!
    '(keycast-command :inherit doom-modeline-debug
                      :height 0.9)
    '(keycast-key :inherit custom-modified
                  :height 1.1
                  :weight bold)))
